### PR TITLE
Validate input types for `torch.nn.Linear` and `torch.nn.Bilinear`

### LINKED ISSUE #135463: Setting a wrong type value instead of a correct integer typed value to `in_features` or `out_features` argument of `nn.Linear()` gets indirect error messages
### 🐛 Describe the bug

Setting a wrong type value instead of a correct `float` type value to `in_features` and `out_features` argument of [nn.Linear()](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) gets the indirect error messages as shown below:

```python
import torch
from torch import nn

my_tensor = torch.tensor([2., 7., 4.])
                             # ↓↓
linear = nn.Linear(in_features=3., out_features=5)
```

> TypeError: empty(): argument 'size' failed to unpack the object at pos 2 with error "type must be tuple of ints,but got float"

```python
import torch
from torch import nn

my_tensor = torch.tensor([2., 7., 4.])
                                             # ↓↓↓↓↓↓
linear = nn.Linear(in_features=3, out_features=5.+0.j)
```

```
TypeError: empty() received an invalid combination of arguments - got (tuple, dtype=NoneType, device=NoneType), but expected one of:
 * (tuple of ints size, *, tuple of names names, torch.memory_format memory_format = None, torch.dtype dtype = None, torch.layout layout = None, torch.device device = None, bool pin_memory = False, bool requires_grad = False)
 * (tuple of ints size, *, torch.memory_format memory_format = None, Tensor out = None, torch.dtype dtype = None, torch.layout layout = None, torch.device device = None, bool pin_memory = False, bool requires_grad = False)
```

So, the error messages should be something like as shown below:

> `in_features` must be `float`.

> `out_features` must be `float`.

### Versions

```python
import torch

torch.__version__ # 2.4.0+cu121
```

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @malfet

### PR DESCRIPTION
**Pull request by sanskarmodi8**
Adding validation checks to check the input types and display better error messages for the same.
Fixes https://github.com/pytorch/pytorch/issues/135463 

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @malfet

### CODE REVIEW

**Review by malfet — CHANGES_REQUESTED** [review]
Please add unit tests and restrict your modifications to specific module, as this PR prevents anyone from extending the module and defining in_features as say floating point argument

**Comment by sanskarmodi8** [comment]
> Please add unit tests and restrict your modifications to specific module, as this PR prevents anyone from extending the module and defining in_features as say floating point argument

Ok. I am working on these changes

**Comment by sanskarmodi8** [comment]
Hi @malfet, 
I have made my changes specific to Linear and Bilinear modle and implemented unit tests for the same as you requested.

File: test/nn/test_linear_module_validation.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import unittest
+import torch
+from torch.nn import Parameter
+from torch.nn.modules.linear import Linear, Bilinear
+
+class TestLinearValidation(unittest.TestCase):
```
Comment (mikaylagawarecki) [inline test/nn/test_linear_module_validation.py:6]: these can be added as [`ErrorModuleInput`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_modules.py#L188)s for `Linear` and `Bilinear`, [see usage here for an example](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_modules.py#L4215)

**Review by mikaylagawarecki — COMMENTED** [review]
change looks good, just need to move the test

File: test/nn/test_linear_module_validation.py (hunk 1 above)
Comment (sanskarmodi8) [reply test/nn/test_linear_module_validation.py]: Thans for the suggestion. I haved added them accordingly

**Comment by mikaylagawarecki** [comment]
looks like a lot of lines were changed due to your linter, can you undo your linting and use the pytorch lintrunner

`make setup_lint`, `make quicklint` from your pytorch folder

**Comment by sanskarmodi8** [comment]
> looks like a lot of lines were changed due to your linter, can you undo your linting and use the pytorch lintrunner
> 
> `make setup_lint`, `make quicklint` from your pytorch folder

Sure. I'm working on it

**Review by malfet — DISMISSED** [review]
LGTM, thank you for the fix

**Comment by sanskarmodi8** [comment]
> @pytorchbot merge -f "Let's land those right away"

Hey @malfet ,
Although the PR is merged, it's not showing in the Pull Requests section in merged filter. It appears only in closed filter.

**Comment by malfet** [comment]
> Is the PR is merged, it's not showing in the Pull Requests section in merged filter. It appears only in closed filter. !

It was not merged, but it's part of the trunk, see https://github.com/pytorch/pytorch/commit/e157ce3ebbb3f30d008c15914e82eb74217562f0

**Comment by sanskarmodi8** [comment]
> filter

Oh, ok

**Comment by sanskarmodi8** [comment]
> > Is the PR is merged, it's not showing in the Pull Requests section in merged filter. It appears only in closed filter. !
> 
> It was not merged, but it's part of the trunk, see https://github.com/pytorch/pytorch/commit/e157ce3ebbb3f30d008c15914e82eb74217562f0

Oh, ok.

**Comment by malfet** [comment]
I.e. things like that should work:
```
>>> torch.nn.Linear(np.array([1, 2, 3])[0], 2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/nshulga/git/pytorch/pytorch/torch/nn/modules/linear.py", line 103, in __init__
    raise TypeError(f"Expected int for in_features but got {type(in_features)}")
TypeError: Expected int for in_features but got <class 'numpy.int64'>
>>> 
```

**Comment by sanskarmodi8** [comment]
> I.e. things like that should work:
> ```
> >>> torch.nn.Linear(np.array([1, 2, 3])[0], 2)
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/Users/nshulga/git/pytorch/pytorch/torch/nn/modules/linear.py", line 103, in __init__
>     raise TypeError(f"Expected int for in_features but got {type(in_features)}")
> TypeError: Expected int for in_features but got <class 'numpy.int64'>
> >>> 
> ```

Thanks for feedback. I'm working on it.

**Pull request closed without merging**